packages/app/app/actions/queue.ts

```diff
@@ -262,11 +264,23 @@ export function addPlaylistTracksToQueue(tracks) {
 function dispatchWithShuffle(dispatch, getState, action) {
   const state = getState();
   const settings = state.settings;
-  const queue = state.queue;
+  let queue = state.queue;
 
   if (settings.shuffleQueue) {
-    const index = _.random(0, queue.queueItems.length - 1);
-    dispatch(selectSong(index));
+    let unplayedQueue = queue.queueItems.filter(item => !item.played);
+    if (unplayedQueue.length === 0){
+      queue.queueItems.map(item => dispatch(updateQueueItem({ ... item, played: false})));
```

Comment:
For performance reasons, this should not result in an action being dispatched for every item in the queue.

Reply:
this will actually not happen as currently the app will keep adding new songs to the queue. Even if this happens it will be only after a user has played every song in the queue. Do you still think this is an efficiency problem and any advice on what to do with this?

Reply:
New songs are only added if you have the autoradio setting enabled. We could introduce a new action that would reset this flag on the entire queue at once, so we would only dispatch one action in this case.